scylla-cql/src/frame/request/mod.rs

```diff
@@ -48,3 +48,137 @@ pub trait Request {
 pub trait DeserializableRequest: Request + Sized {
     fn deserialize(buf: &mut &[u8]) -> Result<Self, ParseError>;
 }
+
+#[cfg(test)]
+mod tests {
+    use std::borrow::Borrow;
+
+    use bytes::Bytes;
+
+    use crate::{
+        frame::{
+            request::{
+                batch::{BatchOwned, BatchType},
+                execute::ExecuteOwned,
+                query::{QueryOwned, QueryParametersOwned},
+                DeserializableRequest, Request,
+            },
+            types::SerialConsistency,
+            value::SerializedValues,
+        },
+        Consistency,
+    };
+
+    use super::{batch::BatchStatementOwned, query::QueryParametersOwnershipGeneric};
+
+    impl<Values: Borrow<SerializedValues> + PartialEq> PartialEq
+        for QueryParametersOwnershipGeneric<Values>
+    {
+        fn eq(&self, other: &Self) -> bool {
+            self.consistency == other.consistency
+                && self.serial_consistency == other.serial_consistency
+                && self.timestamp == other.timestamp
+                && self.page_size == other.page_size
+                && self.paging_state == other.paging_state
+                && self.values == other.values
+        }
+    }
+
+    impl PartialEq for BatchOwned {
+        fn eq(&self, other: &Self) -> bool {
+            self.statements
+                .clone()
+                .zip(other.statements.clone())
+                .all(|(fst, snd)| fst == snd)
+                && self.statements_count == other.statements_count
+                && self.batch_type == other.batch_type
+                && self.consistency == other.consistency
+                && self.serial_consistency == other.serial_consistency
+                && self.timestamp == other.timestamp
+                && self.values == other.values
+        }
+    }
```

Comment:
As far as I understand, `#[derive()]` does not support deriving implementations only for some specialisations of generic type (in this case, specialisations whose `Values` type parameter implements `PartialEq`).

Reply:
Did you try it? Derive macros operate on the whole definition of a struct, so in theory it could take the trait bound into account.

If not, then maybe we could drop this bound?

Alternatively, if we use `Cow` then the problem should solve itself.

Reply:
Derive macro means: "implement the trait automatically. Dot.". It does not say: "implement the trait for all generic parameters that implement the trait as well", and does not allow for this.

Reply:
A derive macro has access to the whole definition of a struct, including the generic parameters and their trait bounds. Such a macro can copy those bounds into the generated `impl` block. For example, the derive macro I implemented [here](https://github.com/scylladb/scylla-rust-driver/pull/665/commits/c6e08762a8476b4e938b16152fb65941a21a5ae0) takes the trait bounds of each parameter into account. Therefore, it is allowed. The question is just whether the stock derive macros do this.

Reply:
Can we change `statements` not to be an iterator? Can it be `Cow<'x, [T]>` where `T: SomeTraitThatConvertsToBatchFromScyllaCql`?

Reply:
Yes! I don't believe it after the lifetime storm I had to overcome, but I managed to accomplish that.